### PR TITLE
petsc: add fotran build argument

### PR DESCRIPTION
**Pull request by wucke13**
###### Motivation for this change

I stumbled up on some piece of software which needs this, so I added it quickly.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/petsc/default.nix [hunk 1]
```diff
@@ -2,27 +2,31 @@
...
+    sha256 = "0yf5lm8dqpbz1yq9nvd5g2iy2w0rgjp9vcj44ba63rwig37j3clp";
   };
 
-  nativeBuildInputs = [ blas gfortran.cc.lib lapack python ];
+  nativeBuildInputs = [ blas gfortran gfortran.cc.lib lapack python ];
 
-  prePatch = stdenv.lib.optionalString stdenv.isDarwin ''
+  prePatch = ''
+    substituteInPlace configure \
+      --replace /bin/sh /usr/bin/python
```
Comment (bhipple) [inline pkgs/development/libraries/science/math/petsc/default.nix:16]: Did you really mean to replace the shell with python?

Comment (wucke13) [reply pkgs/development/libraries/science/math/petsc/default.nix:16]: Hihi, yes, I do. If upstream does weird things, we have to keep pace, I guess. Here is the current version of the configure script:

```
#!/bin/sh

"true" '''\'
for pyimpl in python3 python python2; do
  which $pyimpl > /dev/null 2>&1  && echo "Using $pyimpl for Python" && exec $pyimpl "$0" "$@"
done
echo "Error! Could not locate python! Please install and have python3, python, or python2 in PATH."
echo "Or use: /path/to/valid/python configure $@"
exit 127
'''

import sys, os

if sys.version_info < (2,6) or (sys.version_info >= (3,0) and sys.version_info < (3,4)):
  print('************************************************************************')
  print('*      Python version 2.6+ or 3.4+ is required to run ./configure      *')
  print('*         Try: "python2.7 ./configure" or "python3 ./configure"        *')
  print('************************************************************************')
  sys.exit(4)

sys.path.insert(0, os.path.abspath('config'))
import configure
configure.petsc_configure([])
````

TL;DR Upstream does some ugly she-py-bang stuff to find a suitable python executor, my change streamlines this process.

Comment (bhipple) [reply pkgs/development/libraries/science/math/petsc/default.nix:16]: Oh my ... worth a comment inline in the nix expr w/ your TLDR and/or a link, for future us

Comment (wucke13) [reply pkgs/development/libraries/science/math/petsc/default.nix:22]: Sure, I added it.
